lib/srv/desktop/audit.go

```diff
@@ -29,487 +30,465 @@ import (
 	"github.com/gravitational/teleport/lib/tlsca"
 )
 
-func (s *WindowsService) onSessionStart(ctx context.Context, recorder libevents.SessionPreparerRecorder, id *tlsca.Identity, startTime time.Time, windowsUser, sessionID string, desktop types.WindowsDesktop, err error) {
-	userMetadata := id.GetUserMetadata()
-	userMetadata.Login = windowsUser
+// desktopSessionAuditor is used to build session-related events
+// which are emitted to Teleport's audit log
+type desktopSessionAuditor struct {
+	clock clockwork.Clock
+
+	sessionID   string
+	identity    *tlsca.Identity
+	windowsUser string
+	desktop     types.WindowsDesktop
+
+	startTime          time.Time
+	clusterName        string
+	desktopServiceUUID string
+
+	auditCache sharedDirectoryAuditCache
+}
+
+func (s *WindowsService) newSessionAuditor(
+	sessionID string,
+	identity *tlsca.Identity,
+	windowsUser string,
+	desktop types.WindowsDesktop) *desktopSessionAuditor {
+	return &desktopSessionAuditor{
+		clock: s.cfg.Clock,
+
+		sessionID:   sessionID,
+		identity:    identity,
+		windowsUser: windowsUser,
+		desktop:     desktop,
+
+		startTime:          s.cfg.Clock.Now().UTC().Round(time.Millisecond),
+		clusterName:        s.clusterName,
+		desktopServiceUUID: s.cfg.Heartbeat.HostUUID,
+
+		auditCache: newSharedDirectoryAuditCache(),
+	}
+}
+
+func (d *desktopSessionAuditor) makeSessionStart(err error) *events.WindowsDesktopSessionStart {
```

Comment:
Convention: methods starting with `make` are just constructors - they always return an event.

Methods starting with `on` may or may not return an event (some of these just update the internal state for future events).